<?xml version="1.0"?>
<robot xmlns:xacro="http://www.ros.org/wiki/xacro" name="link_4">

<!-- GEOMETRY -->
    <xacro:macro name="link_4_geometry">
        <origin rpy="0 0 0" xyz="${link_4_length/2} 0 0" />
        <geometry>
            <box size="${link_4_length} ${link_4_width} ${link_4_height}"/>
        </geometry>
    </xacro:macro>

    <!-- VISUAL -->
    <xacro:macro name="link_4_visual">
        <origin xyz="0 0 0" rpy="0 0 0"/>
        <geometry>
            <mesh filename="package://phobos_meshes/rover_stl/chwytak.STL" scale="0.001 0.001 0.001"/>
        </geometry>
    </xacro:macro>


<!-- MATERIAL -->
    <xacro:macro name="link_4_material">
        <material name="link_4_color">
            <color rgba=".4 .4 .8 1"/>
        </material>
    </xacro:macro>

<!-- LINK -->
    <xacro:macro name="link_4" params="name parent x y z">

        <link name="${name}">
            <visual>
                <xacro:link_4_visual/>
                <xacro:link_4_material/>
            </visual>
            <collision>
                <xacro:link_4_geometry/>
            </collision>
            <inertial>
                <origin rpy="0 0 0" xyz="${link_4_length/2} 0 ${link_4_height/2}" />
                <mass value="0.7"/>
                <inertia ixx=".003" ixy="0.000" ixz="0.000" iyy=".003" iyz="0.000" izz=".003"/>
            </inertial>
        </link>
        <gazebo reference="${name}">
            <mu1> .2 </mu1>
            <mu2> .2 </mu2>
            <material> Gazebo/Grey </material>
            <selfCollide>True</selfCollide>
        </gazebo>

        <joint name="${name}_joint" type="revolute">
            <parent link="${parent}"/>
            <child link="${name}"/>
            <origin rpy="0 0 0" xyz="${x} ${y} ${z}"/>
            <axis xyz="1 0 0"/>
            <limit effort="40" velocity="3" lower="-3.14" upper="3.14"/>
            <dynamics damping="1"/>
        </joint>
        <!--gazebo reference="${name}_joint">
            <stopCfm>0.8</stopCfm>
            <stopErp>0.5</stopErp>
            <provideFeedback>true</provideFeedback>
            <implicitSpringDamper>true</implicitSpringDamper>
            <fudgeFactor>0.0</fudgeFactor>
        </gazebo-->
        <gazebo>
            <plugin name="joint_limiter_${name}" filename="libJointLimiter.so">
                <joint>${name}_joint</joint>
                <topic>/simulation/controller/vel/${name}</topic>
                <lower>-3.14</lower>
                <upper>3.14</upper>
            </plugin>
        </gazebo>
        <gazebo>
            <plugin name="joint_encoder_${name}" filename="libJointEncoder.so">
                <joint>${name}_joint</joint>
                <topic>/simulation/encoder/abs/${name}</topic>
                <lower>-3.14</lower>
                <upper>3.14</upper>
                <absolute>1</absolute>
                <resolution>10000</resolution>
            </plugin>
        </gazebo>
        <gazebo>
            <plugin name="joint_controller_${name}" filename="libControllerPID.so">
                <joint>${name}_joint</joint>
                <topic>/simulation/controller/vel/${name}</topic>
                <Kp>1.05</Kp>
                <Ki>0</Ki>
                <Kd>0.02</Kd>
            </plugin>
        </gazebo>


        <gazebo reference="${name}">
            <mu1> .2 </mu1>
            <mu2> .2 </mu2>
            <material> Gazebo/Yellow </material>
            <selfCollide>True</selfCollide>
            <!--kp>1.0</kp>
            <kd>1.0</kd>
            <stopCfm>0.8</stopCfm>
            <stopErp>0.5</stopErp>
            <provideFeedback>true</provideFeedback>
            <cfmDamping>true</cfmDamping>
            <fudgeFactor>0.0</fudgeFactor-->
        </gazebo>

        <!--transmission name="${name}_transmission">
            <type>transmission_interface/SimpleTransmission</type>
            <joint name="${name}_joint">
                <hardwareInterface>EffortJointInterface</hardwareInterface>
            </joint>
            <actuator name="${name}_motor">
                <hardwareInterface>EffortJointInterface</hardwareInterface>
                <mechanicalReduction>1</mechanicalReduction>
            </actuator>
        </transmission-->

    </xacro:macro>
</robot>
